flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -819,6 +823,8 @@ else if (transitionState(current, ExecutionState.FAILED, t)) {
 		}
 		finally {
 			try {
+				waitInvokableCancelCompletion();
```

Comment:
Could you add a test for clean shutdown case as well? (to cover for the bug in the previous version that I commented in the PR)

Reply:
This case is already covered by `TaskTest.testTerminationFutureCompletesOnNormalExecution`.
I'll add a bit waiting and re-check there.